crates/sui-types/src/temporary_store.rs

```diff
@@ -653,9 +676,64 @@ impl<S> TemporaryStore<S> {
             } else {
                 Some(inner.events.digest())
             },
-            transaction_dependencies,
+            transaction_dependencies.clone(),
         );
-        (inner, effects)
+
+        let mut effects_check_error = Ok(());
+
+        // We might have to check the size of effects and rollback if too large
+        if enforce_effects_size_limit {
+            // Check the effects size and reset if too large
+            let size_check_status = match bcs::serialized_size(&effects) {
+                Err(_e) => Err(ExecutionError::new_with_source(
+                    ExecutionErrorKind::InvariantViolation,
+                    "Transaction effects should not fail serialization",
+                )),
+                Ok(serialized_effects_size) => {
+                    let max_serialized_effects_size =
+                        protocol_config.max_serialized_tx_effects_size_bytes();
+                    if serialized_effects_size > max_serialized_effects_size {
+                        Err(ExecutionError::new_with_source(
+                            ExecutionErrorKind::EffectsTooLarge {
+                                current_size: serialized_effects_size,
+                                max_size: max_serialized_effects_size,
+                            },
+                            "Transaction effects are too large",
+                        ))
+                    } else {
+                        Ok(())
+                    }
+                }
+            };
+
+            // Rollback if size too large. Only gas smashed must remain in effects
+            // TODO (Ade): should we surcharge for the work required to rollback?
+            if let Err(size_check_err) = size_check_status {
+                // Reset temp store, but exclude gas objects, which must remain for charging
+                self.reset(gas.iter().map(|(q, _, _)| *q).collect());
```

Comment:
wonder if it is better to retain or reapply but we'll sort it out....

Reply:
Good point, that approach seems cleaner. That was my initial approach but it meant having to handle ensuring deducting gas all over again among other things. And gas deduction is already done much earlier.
But I'm okay either way.

Reply:
I think it would be a lot cleaner to just reapply the write for the gas object, which should be easy to do?

Reply:
One sec. Gonna try something...

Reply:
This has been addressed